crates/wasm-mutate/src/mutators/peephole/swap_commutative.rs

```diff
@@ -0,0 +1,176 @@
+use rand::{prelude::SmallRng, Rng};
+use wasm_encoder::{CodeSection, Function, Instruction, Module, ValType};
+use wasmparser::{CodeSectionReader, FunctionBody, Operator, SectionReader};
+
+use crate::{error::EitherType, module::*, Error, ModuleInfo, Result, WasmMutate};
+
+use super::CodeMutator;
+
+pub struct SwapCommutativeOperator;
+
+impl SwapCommutativeOperator {
+    fn is_commutative(&self, op: &Operator) -> bool {
+        match op {
+              Operator::I32Add | Operator::I32Mul | Operator::I32Or | Operator::I32And | Operator::I32Xor
+            | Operator::I64Add | Operator::I64Mul | Operator::I64And | Operator::I64Or | Operator::I64Xor
+            | Operator::F32Add | Operator::F32Mul // Check for float incosistency 
+            | Operator::F64Add | Operator::F64Mul // Check for float incosistency
+            // TODO do the others
+            => {
+                true
+            }
+            _ => {
+                false
+            }
+        }
+    }
+
+    fn get_operator_type(&self, op: &Operator) -> Result<ValType> {
+        match op {
+            Operator::I32Add
+            | Operator::I32Mul
+            | Operator::I32Or
+            | Operator::I32And
+            | Operator::I32Xor => Ok(ValType::I32),
+            Operator::I64Add
+            | Operator::I64Mul
+            | Operator::I64And
+            | Operator::I64Or
+            | Operator::I64Xor => Ok(ValType::I64),
+            Operator::F32Add | Operator::F32Mul => Ok(ValType::F32),
+            Operator::F64Add | Operator::F64Mul => Ok(ValType::F64),
+            // TODO do the others
+            _ => Err(Error::UnsupportedType(EitherType::Operator(format!(
+                "{:?}",
+                op
+            )))),
+        }
+    }
+}
+
+impl CodeMutator for SwapCommutativeOperator {
+    fn mutate(
+        &self,
+        _: &WasmMutate,
+        _: &mut SmallRng,
+        funcreader: &mut FunctionBody,
+        operator_index: usize,
+        input_wasm: &[u8],
+    ) -> Result<Function> {
+        let opreader = funcreader.get_operators_reader().unwrap();
+        let body_range = opreader.range();
+        let opreaderiterator = opreader.into_iter_with_offsets();
+        let operators = opreaderiterator
+            .collect::<wasmparser::Result<Vec<(Operator, usize)>>>()
+            .unwrap();
```

Comment:
Haven't we already parsed the operators by now? E.g. the operators are passed to the `can_mutate` method.

Can we pass them in here as well? And the locals and whatever else is needed. This way we don't need to parse the same thing twice.

Reply:
Thats right, but the thing is that I dont want to map each wasmparser operator to wasm-encoder instruction. If we pass the operators vector, we maybe can pass the offset inside the byte stream as well as a vector of tuples (operator, offset). With  the mutation position, the previous byte stream could be copied with this offset, we do the mutation and the we copy the byte stream from the next operator offset. 
I am working in this, I will come with a better solution :)

Reply:
Ok, now we pass the operators and their respective offset inside the byte stream, besides the function reader, we might need to use locals_reader